core/src/main/java/org/infinispan/protostream/annotations/impl/GeneratedMarshallerBase.java

```diff
@@ -46,30 +44,12 @@ protected final <T> void writeNestedMessage(BaseMarshallerDelegate<T> marshaller
       if (ctx.depth() >= maxNestedMessageDepth) {
          throw log.maxNestedMessageDepth(maxNestedMessageDepth, message.getClass());
       }
-      try (NestedWriter nested = new NestedWriter(ctx, fieldNumber)) {
-         writeMessage(marshallerDelegate, nested.writer, message);
-      }
-   }
-
-   public static class NestedWriter implements AutoCloseable {
-      final RandomAccessOutputStream baos;
-      final TagWriterImpl writer;
-      private final ProtobufTagMarshaller.WriteContext ctx;
-      private final int fieldNumber;
+      if (!(ctx instanceof TagWriter)) {
+         throw new IllegalArgumentException("Context [" + ctx + "] must be an implementation of TagWriter");
 
-      public NestedWriter(ProtobufTagMarshaller.WriteContext ctx, int fieldNumber) {
-         this.ctx = ctx;
-         this.fieldNumber = fieldNumber;
-         this.baos = new RandomAccessOutputStreamImpl();
-         this.writer = TagWriterImpl.newNestedInstance(ctx, baos);
       }
-
-      public TagWriterImpl getWriter() {
-         return writer;
-      }
-
-      public void close() throws IOException {
-         ctx.getWriter().writeBytes(fieldNumber, baos.getByteBuffer());
+      try (TagWriter nestedWriter = ((TagWriter) ctx).subWriter(fieldNumber, true)) {
+         marshallerDelegate.marshall((ProtobufTagMarshaller.WriteContext) nestedWriter, null, message);
```

Comment:
`ctx`does not need to implement `TagWriter`. You can do
```java
      try (TagWriter nestedWriter = ctx.getWriter().subWriter(fieldNumber, true)) {
         marshallerDelegate.marshall((ProtobufTagMarshaller.WriteContext) nestedWriter, null, message);
      }
```

Reply:
Ah, I didn't notice that before 👍